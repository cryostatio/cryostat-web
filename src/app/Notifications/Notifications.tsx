/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { BehaviorSubject, Observable } from 'rxjs';
import { concatMap, filter, first, map } from 'rxjs/operators';
import { AlertVariant } from '@patternfly/react-core';
import { nanoid } from 'nanoid';
import { NotificationCategory } from '@app/Shared/Services/NotificationChannel.service';

export interface Notification {
  hidden?: boolean;
  read?: boolean;
  key?: string;
  title: string;
  message?: string | Error;
  category?: string;
  variant: AlertVariant;
  timestamp?: number;
}

export class Notifications {
  private readonly _notifications$: BehaviorSubject<Notification[]> = new BehaviorSubject<Notification[]>([]);
  private readonly _drawerState$: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);

  constructor() {
    this._drawerState$
      .pipe(
        filter((v) => v),
        concatMap(() => this._notifications$.pipe(first()))
      )
      .subscribe((prev) =>
        this._notifications$.next(
          prev.map((n) => ({
            ...n,
            hidden: true,
          }))
        )
      );
  }

  drawerState(): Observable<boolean> {
    return this._drawerState$.asObservable();
  }

  setDrawerState(state: boolean): void {
    this._drawerState$.next(state);
  }

  notify(notification: Notification): void {
    if (!notification.key) {
      notification.key = nanoid();
    }
    notification.read = false;
    if (notification.hidden === undefined) {
      notification.hidden = this._drawerState$.getValue();
    }
    notification.timestamp = +Date.now();
    if (notification.message instanceof Error) {
      notification.message = JSON.stringify(notification.message, Object.getOwnPropertyNames(notification.message));
    } else if (typeof notification.message !== 'string') {
      notification.message = JSON.stringify(notification.message);
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      prev.unshift(notification);
      this._notifications$.next(prev);
    });
  }

  success(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.success, hidden });
  }

  info(title: string, message?: string | Error, category?: string, hidden?: boolean): void {
    this.notify({ title, message, category, variant: AlertVariant.info, hidden });
  }

  warning(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.warning });
  }

  danger(title: string, message?: string | Error, category?: string): void {
    this.notify({ title, message, category, variant: AlertVariant.danger });
  }

  notifications(): Observable<Notification[]> {
    return this._notifications$.asObservable();
  }

  unreadNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => !n.read)));
  }

  actionsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter((n) => this.isActionNotification(n))));
  }

  cryostatStatusNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(
      map((a) =>
        a.filter(
          (n) => (this.isWsClientActivity(n) || this.isJvmDiscovery(n)) && !Notifications.isProblemNotification(n)
        )
      )
    );
  }

  problemsNotifications(): Observable<Notification[]> {
    return this.notifications().pipe(map((a) => a.filter(Notifications.isProblemNotification)));
  }

  setHidden(key?: string, hidden: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.hidden = hidden;
        }
      }
      this._notifications$.next(prev);
    });
  }

  setRead(key?: string, read: boolean = true): void {
    if (!key) {
      return;
    }
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        if (n.key === key) {
          n.read = read;
        }
      }
      this._notifications$.next(prev);
    });
  }

  markAllRead(): void {
    this._notifications$.pipe(first()).subscribe((prev) => {
      for (let n of prev) {
        n.read = true;
      }
      this._notifications$.next(prev);
    });
  }

  clearAll(): void {
    this._notifications$.next([]);
  }

  private isActionNotification(n: Notification): boolean {
    return !this.isWsClientActivity(n) && !this.isJvmDiscovery(n) && !Notifications.isProblemNotification(n);
  }

  private isWsClientActivity(n: Notification): boolean {
    return n.category === NotificationCategory.WsClientActivity;
  }

  private isJvmDiscovery(n: Notification): boolean {
    return n.category === NotificationCategory.TargetJvmDiscovery;
  }

  static isProblemNotification(n: Notification): boolean {
    return n.variant === AlertVariant.warning || n.variant === AlertVariant.danger;
  }
}

const NotificationsInstance = new Notifications();

const NotificationsContext = React.createContext(NotificationsInstance);

export { NotificationsContext, NotificationsInstance };
