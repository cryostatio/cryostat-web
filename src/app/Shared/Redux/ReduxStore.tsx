/*
 * Copyright The Cryostat Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { combineReducers, configureStore, PreloadedState } from '@reduxjs/toolkit';
import dashboardConfigReducer, * as dashboardConfigSlice from './Configurations/DashboardConfigSlice';
import topologyConfigReducer, * as topologyConfigSlice from './Configurations/TopologyConfigSlice';
import automatedAnalysisFilterReducer, * as automatedAnalysisFilterSlice from './Filters/AutomatedAnalysisFilterSlice';
import recordingFilterReducer, * as recordingFilterSlice from './Filters/RecordingFilterSlice';
import topologyFilterReducer, * as topologyFilterSlice from './Filters/TopologyFilterSlice';
import { persistMiddleware } from './Middlewares/PersistMiddleware';

// Export actions
export const {
  dashboardConfigAddCardIntent,
  dashboardConfigDeleteCardIntent,
  dashboardConfigReorderCardIntent,
  dashboardConfigResizeCardIntent,
  dashboardConfigFirstRunIntent,
  dashboardConfigCreateLayoutIntent,
  dashboardConfigDeleteLayoutIntent,
  dashboardConfigRenameLayoutIntent,
  dashboardConfigReplaceLayoutIntent,
  dashboardConfigFavoriteLayoutIntent,
  dashboardConfigCreateTemplateIntent,
  dashboardConfigDeleteTemplateIntent,
  dashboardConfigClearLayoutIntent,
  dashboardConfigTemplateHistoryPushIntent,
  dashboardConfigTemplateHistoryClearIntent,
  defaultDashboardConfigs,
} = dashboardConfigSlice;
export const {
  recordingAddFilterIntent,
  recordingDeleteFilterIntent,
  recordingAddTargetIntent,
  recordingDeleteTargetIntent,
  recordingDeleteCategoryFiltersIntent,
  recordingUpdateCategoryIntent,
  recordingDeleteAllFiltersIntent,
  defaultRecordingFilters,
} = recordingFilterSlice;
export const {
  automatedAnalysisAddGlobalFilterIntent,
  automatedAnalysisAddFilterIntent,
  automatedAnalysisAddTargetIntent,
  automatedAnalysisDeleteAllFiltersIntent,
  automatedAnalysisDeleteCategoryFiltersIntent,
  automatedAnalysisDeleteFilterIntent,
  automatedAnalysisDeleteTargetIntent,
  automatedAnalysisUpdateCategoryIntent,
  defaultAutomatedAnalysisFilters,
} = automatedAnalysisFilterSlice;

export const { topologyConfigSetViewModeIntent, topologyDisplayOptionsSetIntent, defaultTopologyConfig } =
  topologyConfigSlice;

export const {
  topologyUpdateCategoryTypeIntent,
  topologyUpdateCategoryIntent,
  topologyAddFilterIntent,
  topologyDeleteAllFiltersIntent,
  topologyDeleteCategoryFiltersIntent,
  topologyDeleteFilterIntent,
  defaultTopologyFilters,
} = topologyFilterSlice;

export const rootReducer = combineReducers({
  dashboardConfigs: dashboardConfigReducer,
  recordingFilters: recordingFilterReducer,
  automatedAnalysisFilters: automatedAnalysisFilterReducer,
  topologyConfigs: topologyConfigReducer,
  topologyFilters: topologyFilterReducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) =>
  configureStore({
    reducer: rootReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(persistMiddleware),
  });

export const store = setupStore();

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
export type StateDispatch = typeof store.dispatch;
export type Store = ReturnType<typeof setupStore>;
