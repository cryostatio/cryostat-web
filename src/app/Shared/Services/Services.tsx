/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import * as React from 'react';
import { NotificationsInstance } from '@app/Notifications/Notifications';
import { TargetService, TargetInstance } from './Target.service';
import { TargetsService } from './Targets.service';
import { ApiService } from './Api.service';
import { NotificationChannel } from './NotificationChannel.service';
import { ReportService } from './Report.service';
import { SettingsService } from './Settings.service';
import { LoginService } from './Login.service';
import { JmxCredentials } from './JmxCredentials.service';

export interface Services {
  target: TargetService;
  targets: TargetsService;
  api: ApiService;
  jmxCredentials: JmxCredentials;
  notificationChannel: NotificationChannel;
  reports: ReportService;
  settings: SettingsService;
  login: LoginService;
}

const settings = new SettingsService();
const jmxCredentials = new JmxCredentials(() => api);
const login = new LoginService(TargetInstance, jmxCredentials, settings);
const api = new ApiService(TargetInstance, NotificationsInstance, login);
const notificationChannel = new NotificationChannel(NotificationsInstance, login);
const reports = new ReportService(login, NotificationsInstance);
const targets = new TargetsService(api, NotificationsInstance, login, notificationChannel);

const defaultServices: Services = {
  target: TargetInstance,
  targets,
  api,
  jmxCredentials,
  notificationChannel,
  reports,
  settings,
  login,
};

const ServiceContext: React.Context<Services> = React.createContext(defaultServices);

export { ServiceContext, defaultServices };
