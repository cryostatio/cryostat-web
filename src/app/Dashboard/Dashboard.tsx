/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  CardConfig,
  dashboardCardConfigResizeCardIntent,
} from '@app/Shared/Redux/Configurations/DashboardConfigSlicer';
import { dashboardCardConfigDeleteCardIntent, RootState, StateDispatch } from '@app/Shared/Redux/ReduxStore';
import { FeatureLevel } from '@app/Shared/Services/Settings.service';
import { TargetView } from '@app/TargetView/TargetView';
import { CardActions, CardBody, CardHeader, Grid, GridItem, gridSpans, Text } from '@patternfly/react-core';
import * as React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Observable, of } from 'rxjs';
import { AddCard } from './AddCard';
import { AutomatedAnalysisCardDescriptor } from './AutomatedAnalysis/AutomatedAnalysisCard';
import { DashboardCardActionMenu } from './DashboardCardActionMenu';
import { ResizableCard } from './ResizableCard';

export interface Sized<T> {
  minimum: T;
  default: T;
  maximum: T;
}

export interface DashboardCardSizes {
  span: Sized<gridSpans>;
  height: Sized<number>;
}

export interface DashboardCardDescriptor {
  title: string;
  cardSizes: DashboardCardSizes;
  description: string;
  descriptionFull: JSX.Element | string;
  component: React.FC<any>;
  propControls: PropControl[];
  advancedConfig?: JSX.Element;
}

export interface PropControl {
  name: string;
  key: string;
  description: string;
  kind: 'boolean' | 'number' | 'string' | 'text' | 'select';
  values?: any[] | Observable<any>;
  defaultValue: any;
}

export interface DashboardProps {}

export interface DashboardCardProps {
  dashboardId: number;
  actions?: JSX.Element[];
}

// TODO remove this
const PLACEHOLDER_CARD_SIZE = {
  span: {
    minimum: 1,
    default: 3,
    maximum: 6,
  },
  height: {
    minimum: Number.NaN,
    default: Number.NaN,
    maximum: Number.NaN,
  },
} as DashboardCardSizes;

const PlaceholderCard: React.FunctionComponent<
  {
    title: string;
    message: string;
    count: number;
    toggleswitch: boolean;
    menu: string;
    asyncmenu: string;
    asyncmenu2: string;
  } & DashboardCardProps
> = (props) => {
  return (
    <ResizableCard
      dashboardId={props.dashboardId}
      cardSizes={PLACEHOLDER_CARD_SIZE}
      className="dashboard-card"
      isRounded
    >
      <CardHeader>
        <CardActions>{...props.actions || []}</CardActions>
      </CardHeader>
      <CardBody>
        <Text>title: {props.title}</Text>
        <Text>message: {props.message}</Text>
        <Text>count: {props.count}</Text>
        <Text>toggle: {String(props.toggleswitch)}</Text>
        <Text>menu: {props.menu}</Text>
        <Text>asyncmenu: {props.asyncmenu}</Text>
        <Text>asyncmenus: {props.asyncmenu2}</Text>
      </CardBody>
    </ResizableCard>
  );
};

export const NonePlaceholderCardDescriptor: DashboardCardDescriptor = {
  title: 'None Placeholder',
  cardSizes: PLACEHOLDER_CARD_SIZE,
  description: 'placeholder',
  descriptionFull: 'This is a do-nothing placeholder with no config',
  component: PlaceholderCard,
  propControls: [],
} as DashboardCardDescriptor;

export const AllPlaceholderCardDescriptor: DashboardCardDescriptor = {
  title: 'All Placeholder',
  cardSizes: PLACEHOLDER_CARD_SIZE,
  description: 'placeholder',
  descriptionFull: 'This is a do-nothing placeholder with all the config',
  component: PlaceholderCard,
  propControls: [
    {
      name: 'string',
      key: 'title',
      defaultValue: 'a short text',
      description: 'a string input',
      kind: 'string',
    },
    {
      name: 'text',
      key: 'message',
      defaultValue: 'a long text',
      description: 'a text input',
      kind: 'text',
    },
    {
      name: 'menu select',
      key: 'menu',
      values: ['choices', 'options'],
      defaultValue: '',
      description: 'a selection menu',
      kind: 'select',
    },
    {
      name: 'menu select 2',
      key: 'asyncmenu',
      values: new Observable((subscriber) => {
        let count = 0;
        const id = setInterval(() => {
          if (count > 2) {
            clearInterval(id);
            setTimeout(() => subscriber.error('Timed Out'), 5000);
          }
          subscriber.next(`async ${count++}`);
        }, 1000);
      }),
      defaultValue: '',
      description: 'an async stream selection menu',
      kind: 'select',
    },
    {
      name: 'menu select 3',
      key: 'asyncmenu2',
      values: of(['arr1', 'arr2', 'arr3']),
      defaultValue: '',
      description: 'an async array selection menu',
      kind: 'select',
    },
    {
      name: 'a switch',
      key: 'toggleswitch',
      defaultValue: false,
      description: 'a boolean input',
      kind: 'boolean',
    },
    {
      name: 'numeric spinner input',
      key: 'count',
      defaultValue: 5,
      description: 'a number input',
      kind: 'number',
    },
  ],
  advancedConfig: <Text>This is an advanced configuration component</Text>,
} as DashboardCardDescriptor;

export const getDashboardCards: (featureLevel?: FeatureLevel) => DashboardCardDescriptor[] = (
  featureLevel?: FeatureLevel
) => {
  let cards = [AutomatedAnalysisCardDescriptor];
  if (featureLevel === undefined || featureLevel === FeatureLevel.DEVELOPMENT) {
    cards = cards.concat([NonePlaceholderCardDescriptor, AllPlaceholderCardDescriptor]);
  }
  return cards;
};

export function getConfigByName(name: string): DashboardCardDescriptor {
  for (const choice of getDashboardCards()) {
    if (choice.component.name === name) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${name}`);
}

export function getConfigByTitle(title: string): DashboardCardDescriptor {
  for (const choice of getDashboardCards()) {
    if (choice.title === title) {
      return choice;
    }
  }
  throw new Error(`Unknown card type selection: ${title}`);
}

export const Dashboard: React.FC<DashboardProps> = (_) => {
  const dispatch = useDispatch<StateDispatch>();
  const cardConfigs: CardConfig[] = useSelector((state: RootState) => state.dashboardConfigs.list);

  const handleRemove = React.useCallback(
    (idx: number) => {
      dispatch(dashboardCardConfigDeleteCardIntent(idx));
    },
    [dispatch]
  );

  const handleResetSize = React.useCallback(
    (idx: number) => {
      const defaultSpan = getConfigByName(cardConfigs[idx].name).cardSizes.span.default;
      if (defaultSpan == cardConfigs[idx].span) {
        return;
      }
      dispatch(dashboardCardConfigResizeCardIntent(idx, defaultSpan));
    },
    [dispatch, cardConfigs]
  );

  return (
    <TargetView pageTitle="Dashboard" compactSelect={false}>
      <Grid hasGutter>
        {cardConfigs.map((cfg, idx) => (
          <GridItem span={cfg.span} key={idx}>
            {React.createElement(getConfigByName(cfg.name).component, {
              ...cfg.props,
              dashboardId: idx,
              actions: [
                <DashboardCardActionMenu
                  key={`${cfg.name}-actions`}
                  onRemove={() => handleRemove(idx)}
                  onResetSize={() => handleResetSize(idx)}
                />,
              ],
            })}
          </GridItem>
        ))}
        <GridItem key={cardConfigs.length}>
          <AddCard />
        </GridItem>
      </Grid>
    </TargetView>
  );
};

export default Dashboard;
