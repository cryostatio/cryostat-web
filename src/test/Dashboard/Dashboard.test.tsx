/*
 * Copyright The Cryostat Authors
 *
 * The Universal Permissive License (UPL), Version 1.0
 *
 * Subject to the condition set forth below, permission is hereby granted to any
 * person obtaining a copy of this software, associated documentation and/or data
 * (collectively the "Software"), free of charge and under any and all copyright
 * rights in the Software, and any and all patent rights owned or freely
 * licensable by each licensor hereunder covering either (i) the unmodified
 * Software as contributed to or provided by such licensor, or (ii) the Larger
 * Works (as defined below), to deal in both
 *
 * (a) the Software, and
 * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
 * one is included with the Software (each a "Larger Work" to which the Software
 * is contributed by such licensors),
 *
 * without restriction, including without limitation the rights to copy, create
 * derivative works of, display, perform, and distribute the Software and make,
 * use, sell, offer for sale, import, export, have made, and have sold the
 * Software and the Larger Work(s), and to sublicense the foregoing rights on
 * either these or other terms.
 *
 * This license is subject to the following condition:
 * The above copyright notice and either this complete permission notice or at
 * a minimum a reference to the UPL must be included in all copies or
 * substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { Dashboard } from '@app/Dashboard/Dashboard';
import { act } from 'react-test-renderer';
import renderer from 'react-test-renderer';
import React from 'react';
import { defaultServices, ServiceContext } from '@app/Shared/Services/Services';
import { Target } from '@app/Shared/Services/Target.service';
import { of } from 'rxjs';
import { NotificationsContext, NotificationsInstance } from '@app/Notifications/Notifications';
import { Provider } from 'react-redux';
import { store } from '@app/Shared/Redux/ReduxStore';

const mockFooConnectUrl = 'service:jmx:rmi://someFooUrl';

const mockFooTarget: Target = {
  connectUrl: mockFooConnectUrl,
  alias: 'fooTarget',
  annotations: {
    cryostat: {},
    platform: {},
  },
};

jest.mock('@app/TargetSelect/TargetSelect', () => ({
  TargetSelect: (props) => <div>Target Select</div>,
}));

jest.mock('@app/Dashboard/AutomatedAnalysis/AutomatedAnalysisCard', () => ({
  AutomatedAnalysisCard: (props) => <div>Automated Analysis Card</div>,
}));

jest
  .spyOn(defaultServices.target, 'target')
  .mockReturnValueOnce(of(mockFooTarget)) // renders correctly
  .mockReturnValueOnce(of()) //
  .mockReturnValue(of(mockFooTarget));

describe('<Dashboard />', () => {
  it('renders correctly', async () => {
    let tree;
    await act(async () => {
      tree = renderer.create(
        <ServiceContext.Provider value={defaultServices}>
          <NotificationsContext.Provider value={NotificationsInstance}>
            <Provider store={store}>
              <Dashboard />
            </Provider>
          </NotificationsContext.Provider>
        </ServiceContext.Provider>
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
