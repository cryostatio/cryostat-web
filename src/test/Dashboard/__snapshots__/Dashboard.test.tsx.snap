// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> renders correctly 1`] = `
<section
  className="pf-c-page__main-section"
>
  <nav
    aria-label="Breadcrumb"
    className="pf-c-breadcrumb"
    data-ouia-component-id="OUIA-Generated-Breadcrumb-1"
    data-ouia-component-type="PF4/Breadcrumb"
    data-ouia-safe={true}
  >
    <ol
      className="pf-c-breadcrumb__list"
    >
      <li
        className="pf-c-breadcrumb__item"
      >
        <h1
          className="pf-c-breadcrumb__heading"
        >
          Dashboard
        </h1>
      </li>
    </ol>
  </nav>
  <div
    className="pf-l-stack pf-m-gutter"
  >
    <div
      className="pf-l-stack__item"
    >
      <div
        className="pf-l-grid pf-m-gutter"
      >
        <div
          className="pf-l-grid__item pf-m-12-col"
        >
          <div>
            Target Select
          </div>
        </div>
        <div
          className="pf-l-grid__item"
        >
          <div
            className="pf-l-grid pf-m-gutter"
          >
            <div
              className="pf-l-grid__item pf-m-6-col"
            >
              <div>
                Automated Analysis Card
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
